### PR TITLE
Restructure macro/prim expansion exceptions

### PR DESCRIPTION
**Pull request by clavin-xlnx**
Signed-off-by: Chris Lavin <clavin@xilinx.com>

### CODE REVIEW

**Comment by clavin-xlnx** [comment]
Handling of macro expansion exceptions led to some primitives not being populated in certain situations (`OBUFTDS` in UltraScale when used as part of the  macros `IOBUFDS_INTERMDISABLE`, `IOBUFDS_DIFF_OUT`, or `IOBUFDS_DIFF_OUT_INTERMDISABLE` were not being repopulated with `OBUFTDS_DUAL_BUF` because replacement criteria was not met).  This has led to some [exceptions](https://github.com/Xilinx/RapidWright/blob/e15df3b18cb91a3ce699fd3db4544a2a4f31973e/src/com/xilinx/rapidwright/edif/EDIFNetlist.java#L1579).  This pull request restructures how primitive cell definitions are added so that if the exceptional replacement does not occur, the fallback definition is available.

**Comment by eddieh-xlnx** [comment]
Easy to build a test?

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class TestEDIFNetlist {
+
+    private static final String PART_NAME = "xcvu440-flga2892-2-e";
+    
+    private static final String TEST_MACRO = "IOBUFDS_INTERMDISABLE";
+    
+    private Design createSampleMacroDesign(String macro, Part part) {
+        String designName = TEST_MACRO +"_design";
+        final EDIFNetlist netlist = EDIFTools.createNewNetlist(designName);
+        final Design design = new Design(designName, PART_NAME);
```
Comment (eddieh-xlnx) [inline test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java:23]: Should `PART_NAME` be `Part.getName()`?

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        design.setNetlist(netlist);
+        design.setPartName(part.getName());
```
Comment (eddieh-xlnx) [inline test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java:25]: Is this necessary give you create `Design` with PART_NAME?

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import org.junit.jupiter.api.io.TempDir;
+
+import com.xilinx.rapidwright.design.Design;
+import com.xilinx.rapidwright.device.Part;
+import com.xilinx.rapidwright.device.PartNameTools;
+import com.xilinx.rapidwright.support.CheckOpenFiles;
+
+class TestEDIFNetlist {
+
+    private static final String PART_NAME = "xcvu440-flga2892-2-e";
```
Comment (eddieh-xlnx) [inline test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java:16]: For runtime reasons consider a smaller device?

**Review by eddieh-xlnx — APPROVED** [review]
Just a few minor nitpicks!

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java (hunk 1 above)
Comment (clavin-xlnx) [reply test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java]: Yes

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java (hunk 2 above)
Comment (clavin-xlnx) [reply test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java]: Nope

File: test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java (hunk 3 above)
Comment (clavin-xlnx) [reply test/com/xilinx/rapidwright/edif/TestEDIFNetlist.java]: Switched to xcku040.

**Review by eddieh-xlnx — APPROVED** [review]

**Pull request merged**